Create PaginatedTable component

keithjgrant opened the pull request:
##### SUMMARY
* Adds PaginatedTable, HeaderRow, HeaderCell components to eventually replace PaginatedDataTables for item list layouts
* Converts Org list and Inventories list to use PaginatedTable

Addresses #6189

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

keithjgrant committed:
add patternfly tables; add PaginatedTable

keithjgrant committed:
add HeaderRow component with table sorting

keithjgrant committed:
convert PaginatedTable to function component

keithjgrant committed:
convert InventoryList to use PaginatedTable

AlexSCorey commented on awx/ui_next/src/components/PaginatedTable/PaginatedTable.jsx:
```diff
@@ -0,0 +1,194 @@
+    );
+  }
+
+  const ToolbarPagination = (
```
This seems to be a duplicate of the same component that starts[here](https://github.com/ansible/awx/pull/8729/files#diff-b942f30c50317e77ddaf815b6d7e3bbe508ea5fa74e43318eee5a31050af5f57R135) with a few exceptions.  Could they be combined?

AlexSCorey commented on awx/ui_next/src/screens/Inventory/InventoryList/InventoryList.jsx:
```diff
@@ -188,6 +191,17 @@ function InventoryList({ i18n }) {
           toolbarSearchableKeys={searchableKeys}
           toolbarRelatedSearchableKeys={relatedSearchableKeys}
+          headerRow={
+            <HeaderRow defaultSortKey="name" qsConfig={QS_CONFIG}>
```
Since we will be doing this on each list I wonder if there is some sort of hook or a component we can use to generate the header rows for us.  I'm imagining something where we send it an array or column header names and it returns the correct header cells.

keithjgrant replied:
HeaderRow is a custom component to take care of a bunch of the sorting logic for us already. I could toy with passing the column data as a prop or something.

I also considered exporting the HeaderRow from the item file (e.g. export InventoryHeaderRow from InventoryListItem) so that the column data all lives in the same place instead of keeping it in sync across two files

AlexSCorey commented on awx/ui_next/src/components/PaginatedTable/ActionsTd.jsx:
```diff
@@ -0,0 +1,33 @@
+`;
+
+export default function ActionsTd({ numActions = 1, children }) {
+  const width = numActions * 40;
```
Could you automatically get the number of actions from the number of children?  If so, doing it that way would reduce the number of props being passed in.  Additionally, I have a preference for not using styled components and css html attributes in the same file where possible. I think doing all of our styling in same place/same way is slightly easier to maintain for the future.

keithjgrant replied:
My initial thought was that doesn't work because some rows are missing certain actions... but now that I think about it again, the way it is now they still need to pass an empty div in the missing slots which isn't ideal...

I may need to put an index prop on the children so they can position themselves correctly in the grid

keithjgrant replied:
OK, I reworked this to determine based on the number of children, and use ActionItems as children to render those children to null for actions/buttons where there user doesn't have RBAC permissions

jlmitch5 commented on awx/ui_next/src/components/DataListToolbar/DataListToolbar.jsx:
```diff
@@ -93,9 +93,11 @@ function DataListToolbar({
-          <ToolbarItem>
-            <Sort qsConfig={qsConfig} columns={sortColumns} onSort={onSort} />
-          </ToolbarItem>
+          {sortColumns && (
```
it's nice that we've included this as a fallback until we convert everything and remove sortColumns across the app.

jlmitch5 commented on awx/ui_next/src/screens/Inventory/InventoryList/InventoryList.jsx:
```diff
@@ -188,6 +191,17 @@ function InventoryList({ i18n }) {
           toolbarRelatedSearchableKeys={relatedSearchableKeys}
+          headerRow={
+            <HeaderRow defaultSortKey="name" qsConfig={QS_CONFIG}>
+              <HeaderCell sortKey="name">{i18n._(t`Name`)}</HeaderCell>
```
This is a very nice, declarative interface!

jakemcdermott replied:
> This is a very nice, declarative interface!

@jlmitch5 I agree!

keithjgrant committed:
simplify ActionsTd interface; add ActionItem component

keithjgrant committed:
table header cleanup; fix inventory status labels

keithjgrant committed:
update tests for PaginatedTable lists

keithjgrant committed:
flush out tests for PaginatedTable and related components

jakemcdermott commented:
This is great

* :+1: from me. I don't have any major suggestions  as far as implementation goes. I **really** like how the bulk of the conversion changes are concentrated in the individual `ListItem`.  I think this will be extremely helpful for converting the lists throughout the app. 

* I pulled this down and tested it out a bit (inventories page) and it works well. It's an overall great improvement to the UX

* ~I _did_ see some errors when running the `make ui-devel` build (npm run build). I can't tell for sure but they seem related to module importing (?). I'm wondering if this is the same issue @ jlmitch5 ran into a while back when previewing some unreleased toolbar components?~  cc @keithjgrant @jlmitch5

 **edit:** nevermind on that last part, see comment below:

jakemcdermott commented:
Nevermind - The error I saw was from building on node 10, rebasing, and then trying a rebuild on node 14 without reinstalling deps (the node version bump landed recently)

A full clean & build after rebase fixed things.

jakemcdermott commented:
:wave: @AlexSCorey @jlmitch5 @keithjgrant 

Is this one ready for test?

AlexSCorey reviewed (approved):
All of my concerns have been addressed.

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryList/InventoryListItem.jsx:
```diff
@@ -85,108 +56,84 @@ function InventoryListItem({
+          : i18n._(t`Inventory`)}
+      </Td>
+      <Td key="organization" dataLabel={i18n._(t`Organization`)}>
+        <Link
+          to={`/organizations/${inventory.summary_fields.organization.id}/details`}
+        >
+          {inventory.summary_fields.organization.name}
+        </Link>
```
```suggestion
        <Link
          to={`/organizations/${inventory?.summary_fields?.organization?.id}/details`}
        >
          {inventory?.summary_fields?.organization?.name}
        </Link>
```

@keithjgrant We'll need to handle the intermediate state for list items when they're being deleted. It appears that inventories can lose their reference to the related organization during that time and so we need to avoid referencing the field to avoid an error:

![Screenshot from 2020-12-17 13-04-42](https://user-images.githubusercontent.com/9753817/102526456-ab066980-4069-11eb-8104-97ce1fbcb785.png)

(Found w/ the latest e2e/visual run :tada: ! :tada: !)

keithjgrant committed:
handle null values on inventories pending deletion

Co-authored-by: Jake McDermott <yo@jakemcdermott.me>

jakemcdermott committed:
Regenerate lockfile

jakemcdermott committed:
Address errors from type system

jakemcdermott committed:
Replace chip key with non-array index

Address a lint error (rule: react/no-array-index-key) by using
chip.key prop val instead of an array index for the component key.

jakemcdermott committed:
Use location hook in routed tabs

Fix a unit test that is failing due to history.location
not updating as expected when a routed tab is selected.

jakemcdermott committed:
Resolve most console errors in tests
